### ISSUE TITLE
Add latest version of Google Apps Scripts to GitHub

### ISSUE THREAD

**Issue by aricforrest** [hackforla/website#3641]
### Overview

The Google Apps Scripts used by the website project should be added to the website git repository. This will allow version control and tracking of changes made to the App Scripts over time.

### Action Items

- [x] Obtain view access to wins-form (response) Sheet and gh-requests (Doc) in the admin Google drive share from a technical lead
- [x] Install node.js and npm (if they are not already installed) on the local workstation
- [x] Update your local gh-pages branch with upstream changes, then checkout a new issue branch 
- [x] edit .gitignore to add a line containing the filename `.clasp.json`, then run `git add .gitignore`
- [x] Create a folder called "google-apps-scripts" in the root of the website repository.
- [x] From within the "google-apps-scripts" folder, create a subfolder for each Apps Script project listed below.  In each subfolder, execute the commands `npm install clasp` and `npm install @types/google-apps-script`
  - [x]  wins-form-responses 
  - [x]  gh-requests
- [x] From either of the above subfolders, execute the command `clasp login`, then authenticate to Google Drive and grant permissions to the clasp app
- [x] In each subfolder, clone the script using the command  `clasp clone <ID>` where the script ID can be found on the corresponding Projects Settings page in the Google Drive IDE, then use `git add` to add the following files to the new issue branch: Code.js, appsscript.json, package.json, package-lock.json
  - [x] wins-form-responses
  - [x] gh-requests
- [x] run `git commit`, then push changes to your fork of gh-pages, then create PR in the usual manner. 

### Merge Team - Upon Completion
 - [ ] Release dependency on #4134 and move that issue into the New Issue Approval column with "ready for dev lead" label.
 - [ ] Remove developer's access to files in the Hfla Google Drive Admin folders.  
 - [ ] Regenerate the gh key using the instructions in the Wins Admin Guide.  
  

Example folder structure:
├── google-apps-scripts
│   └── wins-form-responses
│           └── Code.js, appscript.json, package-lock.json, package.json

### Resources

- [Wins Admin Guide](https://docs.google.com/document/d/1j1JTX5XzotgVCVZ91ImvenedPN5-IVjdSKDukq1B9iE/edit?usp=share_link)
- https://github.com/google/clasp

**Comment by ExperimentsInHonesty** [comment]
Before this issue can get ready for milestone it will need a cleanup. Only fields that have content you need should be in the issue, so for instance, the title and an empty dependency are not wanted.

**Comment by blulady** [comment]
Please don't clean up this issue until we have a chance to bring it up with the team on Tuesday October 25

**Comment by aricforrest** [comment]
Title and dependency fields have been removed.

**Comment by blulady** [comment]
Thanks for letting us use this as an example.

**Comment by roslynwythe** [comment]
Availability: all day 11/1/2022, morning 11/2/2022
ETA: 1 day after granted read access to Google Apps Script in the admin Google drive share

**Comment by roslynwythe** [comment]
To understand the flow of 'win' data I looked in the Site Architecture doc and noticed that the Google Apps Script and 'win' submission data was not described.  Is there other documentation?  If not, I suggest that it be added and I would like to contribute to that effort.

**Comment by roslynwythe** [comment]
requested access to Google Apps Script from @kathrynsilvaconway

**Comment by roslynwythe** [comment]
I unassigned this issue, as I was advised that it was not ready.

**Comment by arpitapandya** [comment]
Hey @roslynwythe looks like you moved this issue from `New Issue Approval` column, Have you confirmed since it's not moved to `prioritized backlog` by pm. Please read my message on slack channel

**Comment by arpitapandya** [comment]
I understand that I and Aric discussed that this issue is a priority before working on #2147 but this issue must move to `prioritized backlog` to work on by PM so that you can work. If you have confusion then please chat on slack.

**Comment by arpitapandya** [comment]
@roslynwythe Apologies, I have to un-assign you from this issue, Once it's in `prioritized backlog` hopefully by next Monday, then you're free to assign yourself and do the work.

**Comment by roslynwythe** [comment]
@arpitapandya I apologize, I shouldn't have changed the Project Board setting.  I'll check back on Monday.  Thanks.

**Comment by ExperimentsInHonesty** [comment]
@aricforrest

## Item 1 - FYI
I changed the formatting of your issue so that it conforms with norms (Overview, action items, etc.).

## Item 2 - FYI
I added this at the bottom of the action items
- [ ] Create wiki page which describes where things are, and how they interact and how to maintain them

## Item 3 - Question
Jessica and the PMS went over this issue and don't understand if you are suggesting we use CLASP after the other work is done, or instead of it, given that it promises to convert existing scripts from Google app scripts to code automatically.

> 📁 Structure Code: clasp automatically converts your flat project on [script.google.com](https://script.google.com/) into folders. For example:
> 
> On script.google.com:
>     tests/slides.gs
>    tests/sheets.gs
> locally:
>    tests/
>       slides.js
>       sheets.js

### List feasibility related issues
- [ ] Changes to these Apps Scripts will still need to be manually updated in each respective Google Document/Sheet after changes are made to the version in GitHub.
- [ ] The usage of [Google clasp](https://github.com/google/clasp) could also be considered for Apps Script development

**Comment by aricforrest** [comment]
@ExperimentsInHonesty 

### Item 1 & 2
Thanks for updating the issue! 

### Item 3
When I originally created this issue I didn't have experience with [clasp ](https://github.com/google/clasp) and just saw the need to add the Apps Scripts to GitHub. I was worried at an organization level permissions would be a hurdle using clasp but it appears it simply depends on the same Google Drive permissions we'll already need to grant for anyone working on this issue. 

Since creating this ticket I have taken some time to play with clasp and it's fairly straight forward. Given the ease of using clasp to clone an existing Apps Script locally and the popularity of clasp, I feel it should be included in this issue (and future Google Apps Script processes) but do note it will require the assignee to take some additional steps on their local host as it requires Node.js.

If needed and approve of going the clasp route, I'm happy to update the issue with clasp specific instructions.

**Comment by ExperimentsInHonesty** [comment]
@aricforrest Thanks for the response.  
Please make a new issue for the CLASP evaluation/decision record
Write-up the benefits to our team and the changes it will require, and what the potential downside is, and the merge team/dev leads to review, discuss with you and make a team recommendation.  Then, if it's accepted, rewrite this issue.  

In the meantime, add a dependency label to this issue, and move it to the icebox with that new issue listed as a dependency (on the top of the issue).

**Comment by roslynwythe** [comment]
- The Action items of this issue have been updated to reflect the use of clasp.  For history see #3722 and [DR: Use of CLASP in Google Apps Script Development](https://github.com/hackforla/website/wiki/DR:-Use-of-CLASP-in-Google-Apps-Script-Development)
- The Google Apps Script development process is more complex than before and I have created a separate issue #4134 documenting that process.

**Comment by roslynwythe** [comment]
- New Issue #4134 will be dependent on completion of this issue, and the following issues that were previously dependent on this issue should be updated to depend on #4134
- [ ] #3881 
- [ ] #4035
- [ ] #2505 
- [ ] #2147

**Comment by ExperimentsInHonesty** [comment]
@jdingeman can you give a sign off on this issue or indicate what is missing.

**Comment by jdingeman** [comment]
Hi @roslynwythe, sent you a message on Slack. Just need to clarify this action because of the names of the subfolders is unclear

> - [ ] From within the "google-apps-scripts" folder, create a subfolder for each Apps Script project listed below.  In each subfolder, execute the commands `npm install clasp` and `npm install @types/google-apps-script`
>   - [ ]  wins-form-responses 
>   - [ ]  gh-requests" subfolder

and it should be ready for prioritization.

**Comment by Sebster3** [comment]
ETA: Monday 3/27 EOD
Availability 1-4PM Fri, 6-9PM Mon

**Comment by Sebster3** [comment]
I have a blocker on the following action items:

1. > From either of the above subfolders, execute the command `clasp login`, then authenticate to Google Drive and grant permissions to the clasp app.
  
   I'm able to successfuly log into my personal google account (which has access to the HFLA Drive), however I wasn't asked specifically to provide access to the HFLA Drive—which the action item seems to imply should happen. Is it sufficient for clasp to be authenticated using my google account?

2. > In each subfolder, clone the script using the command `clasp clone <ID>` where the script ID can be found on the corresponding Projects Settings page in the Google Drive IDE, then use git add to add the following files to the new issue branch: Code.js, appsscript.json, package.json, package-lock.json

   I'm not sure where to find the "Project Settings" page or anything resembling a script ID for the two files mentioned in the issue: `wins-form-responses`, `gh-requests`.

There's currently no active branch for this issue on my forked repo as the final action item consists of pushing the changes made in the previous steps.

Thanks in advance for the help!

**Comment by roslynwythe** [comment]
Hi @Sebster3 - 
Thank you for taking up this issue.  I'm not sure why you weren't asked to authorize clasp to access your Google Drive.  Let's see what happens when you run the clone command.  To get the project id, open the production Sheet (or Doc), then click Extensions..Apps Script, then click the Project Settings (gear) icon in the left column.  The ID can be copied from the Project Settings page.

You should be working in a new issue branch created in Action Item #3.  That branch will be pushed to a branch on your forked repo as the final action item.

**Comment by Sebster3** [comment]
Thanks for helping @roslynwythe! It looks like I'll need edit access to `Wins-form (Responses)` in Drive before I can access the Extensions menu. Also, the `gh-requests` file doesn't appear to have an Extensions menu, which could be a follow up problem after we sort out getting proper permissions on the first file.

<img width="560" alt="Screen Shot 2023-03-29 at 11 40 51 AM" src="https://user-images.githubusercontent.com/9600508/228636507-084e83b3-d5e4-4d16-92a3-4a4d94ece0ff.png">

**Comment by roslynwythe** [comment]
> Thanks for helping @roslynwythe! It looks like I'll need edit access to `Wins-form (Responses)` in Drive before I can access the Extensions menu. Also, the `gh-requests` file doesn't appear to have an Extensions menu, which could be a follow up problem after we sort out getting proper permissions on the first file.
> 
> <img alt="Screen Shot 2023-03-29 at 11 40 51 AM" width="560" src="https://user-images.githubusercontent.com/9600508/228636507-084e83b3-d5e4-4d16-92a3-4a4d94ece0ff.png">

Permissions have been granted.  You should now see an Extensions menu on `gh-requests`.  Let me know if you don't.

**Comment by Sebster3** [comment]
Hm, the Extensions menu on `wins form` remains greyed out and there is no Extensions menu available on `gh-requests`.

![Screenshot 2023-04-01 124722](https://user-images.githubusercontent.com/9600508/229311396-2c05a0bc-17d7-450c-af28-ffd0247caac7.png)

I've sent edit requests directly from the files. hopefully that helps.

**Comment by Sebster3** [comment]
Thanks @roslynwythe! The recent set of permissions you provided got everything back on track and a PR is now open :)

https://github.com/hackforla/website/pull/4365
